tough/src/editor.rs

```diff
@@ -0,0 +1,660 @@
+/// the roles using the data provided, and signs the roles.
+#[derive(Debug)]
+pub struct RepositoryEditor {
+    signed_root: SignedRole<Root>,
+
+    new_targets: Option<HashMap<String, Target>>,
+    existing_targets: Option<HashMap<String, Target>>,
+    targets_version: Option<NonZeroU64>,
+    targets_expires: Option<DateTime<Utc>>,
+    targets_extra: Option<HashMap<String, Value>>,
+
+    snapshot_version: Option<NonZeroU64>,
+    snapshot_expires: Option<DateTime<Utc>>,
+    snapshot_extra: Option<HashMap<String, Value>>,
+
+    timestamp_version: Option<NonZeroU64>,
+    timestamp_expires: Option<DateTime<Utc>>,
+    timestamp_extra: Option<HashMap<String, Value>>,
+}
+
+impl RepositoryEditor {
+    /// Create a new, bare `RepositoryEditor`
+    pub fn new<P>(root_path: P) -> Result<RepositoryEditor>
+    where
+        P: AsRef<Path>,
+    {
+        // Read and parse the root.json. Without a good root, it doesn't
+        // make sense to continue
+        let root_path = root_path.as_ref();
+        let root_buf = std::fs::read(root_path).context(error::FileRead { path: root_path })?;
+        let root_buf_len = root_buf.len() as u64;
+        let root = serde_json::from_slice::<Signed<Root>>(&root_buf)
+            .context(error::FileParseJson { path: root_path })?;
+        let mut digest = [0; SHA256_OUTPUT_LEN];
+        digest.copy_from_slice(ring::digest::digest(&SHA256, &root_buf).as_ref());
+
+        let signed_root = SignedRole {
+            signed: root,
+            buffer: root_buf,
+            sha256: digest,
+            length: root_buf_len,
+        };
+
+        Ok(RepositoryEditor {
+            signed_root,
+            new_targets: None,
+            existing_targets: None,
+            targets_version: None,
+            targets_expires: None,
+            targets_extra: None,
+            snapshot_version: None,
+            snapshot_expires: None,
+            snapshot_extra: None,
+            timestamp_version: None,
+            timestamp_expires: None,
+            timestamp_extra: None,
+        })
+    }
+
+    /// Given a `tough::Repository` and the path to a valid root.json, create a
+    /// `RepositoryEditor`. This `RepositoryEditor` will include all of the targets
+    /// and bits of _extra metadata from the roles included. It will not, however,
+    /// include the versions or expirations and the user is expected to set them.
+    pub fn from_repo<T, P>(root_path: P, repo: Repository<'_, T>) -> Result<RepositoryEditor>
+    where
+        P: AsRef<Path>,
+        T: Transport,
+    {
+        let mut editor = RepositoryEditor::new(root_path)?;
+        editor.targets(repo.targets.signed)?;
+        editor.snapshot(repo.snapshot.signed)?;
+        editor.timestamp(repo.timestamp.signed)?;
+
+        Ok(editor)
+    }
+
+    /// Add an existing `Targets` struct to the repository. Any existing
+    /// targets are preserved.
```

Comment:
Not sure I'm understanding this - existing targets aren't preserved, we're fully replacing `self.existing_targets`.  Should this be called something like `set_targets` or `replace_targets`?  Or were you intending to merge the given targets with `existing_targets`?

Reply:
Another option is having `add_targets` and `clear_targets` methods, so a user can work with `Targets` objects if desired (merging, replacing, etc.), and changing the path-based methods below to `add_target(s)_path` if they want to work with paths..?  One use case we've discussed is merging repos, so this distinction seems useful.

Reply:
Good point. I'll clarify the comment as well.

One method I completely missed is a `remove_target` method as well.

Reply:
Oh, wasn't thinking when I answered initially. When a user passes a `Targets` struct to this method, we preserve any `Target` objects in `self.existing_targets`.  When a user adds targets using the `add_target(s)()` methods, they are saved to `self.new_targets`. We combine the two maps when creating the final repository.

I agree that a `clear_targets()` method would be a good idea.

I'm wondering how we best support a `remove_target()` method. The user would need to supply a target filename, since we store this information as `HashMap<filename, Target_object>`

Reply:
Could wait for a use case / request, but I'd assume they probably have a filename/path, so it could build a Target based on that and remove it from existing/new lists...  doesn't seem urgent if we have clear.

Reply:
Agreed. I'll implement `clear_targets()` for now.